Shortcode block

truongwp opened the pull request:
Fixes #14 

Screencast: https://youtu.be/QO-dxmOPfc4

This introduces the new Formidable Modal block in the block editor to show the modal.

garretlaxton commented:
Dropdown fields and a large border around them and you can't see the content.
![image](https://github.com/Strategy11/formidable-modal/assets/16290508/a4dd92d2-b45b-41d8-9041-7cbc4aecad0e)

I can also enter text and select fields inside of the block.
![image](https://github.com/Strategy11/formidable-modal/assets/16290508/a17b63b2-e6f2-4d19-97e8-594876c8b543)

The submit button is also clickable and causes the page to reload.

truongwp commented:
@garretlaxton Can you replicate the dropdown styling issue on the QA site?

For the submit button issue, we'll probably fix in core since it also happens with our form block. I'll ask Mike in today check-in.

garretlaxton commented:
It looks like the dropdown is an issue with the 2021 theme. See here: https://qa.formidableforms.com/truong/wp-admin/post.php?post=105&action=edit

truongwp commented:
@garretlaxton That issue happens with Formidable Form block too. I logged it as a separate issue.

garretlaxton reviewed (approved):
Gotcha, thanks! This works great then!

AbdiTolesa commented on blocks-src/frm-modal-button/index.js:
```diff
@@ -0,0 +1,50 @@
+
+registerBlockType(
+	{
+		name,
```
I think we don't need this line as we are using the rest operator below.

AbdiTolesa commented on blocks-src/frm-modal/block.json:
```diff
@@ -0,0 +1,19 @@
+  "version": "0.1.0",
+  "title": "Formidable Modal",
+  "category": "widgets",
+  "icon": "smiley",
```
We should change this smiley icon, right? 😄 
Here is how it looks in my environment:
<img width="902" alt="image" src="https://github.com/Strategy11/formidable-modal/assets/41271840/30b1e096-ef30-4ed8-9b5b-692d5b035224">

stephywells replied:
Here's the icon from the design that we should use:
https://www.figma.com/file/SXTkDpu190Du7HSk9SAJ6g/Modal-Add-On?type=design&node-id=203-20981&t=W69Q5FIung59ca7c-0

truongwp commented on blocks-src/frm-modal-button/edit.js:
```diff
@@ -0,0 +1,277 @@
+const NEW_TAB_REL = 'noreferrer noopener';
+
+function WidthPanel( { selectedWidth, setAttributes } ) {
+	function handleChange( newWidth ) {
```
There are 2 reasons I want to keep this.
1. This is in a Width setting, there is just the width is changed, so I think we don't need the "Width" in the function name.
2. Almost parts of this file are copied from the WordPress Button block, so I think keeping it as same as possible will help us troubleshoot problems more easily.

truongwp commented on blocks-src/frm-modal-button/edit.js:
```diff
@@ -0,0 +1,277 @@
+		attributes;
+
+	function onToggleOpenInNewTab( value ) {
+		const newLinkTarget = value ? '_blank' : undefined;
```
I think we should keep it as same as the WordPress Button code as I explained in one of the above changes requests.

truongwp commented:
@AbdiTolesa Thank you. I did some cleanup. Please review this again.

AbdiTolesa commented on classes/frmBtsModApp.php:
```diff
@@ -81,50 +235,66 @@ private static function build_shortcode( $atts ) {
+					$modal .= '</div>';
+					$modal .= '</div>';
+					$modal .= '</div>';
+					$modal .= '</div>';
```
This long code might be worth putting into a new function and just calling that function here to get the modal if possible.

AbdiTolesa commented:
@Crabcyborg @truongwp
As a general comment and discussion point on this PR, it looks we are heavily relying on `__experimental ` features and it is [stated in this issue](https://github.com/WordPress/gutenberg/issues/40316) that every new Gutenberg release means potentially breaking changes. They don't recommend it for any serious production usage. I believe we should hardly use them or at least try to find as much alternative solutions as possible to minimise the maintenance burden required as new Gutenberg versions are released.

What are your thoughts on this?

truongwp commented on blocks-src/frm-modal-content/block.json:
```diff
@@ -0,0 +1,77 @@
+      "source": "attribute",
+      "selector": ".modal-dialog",
+      "attribute": "data-size"
+    },
```
I see in the `.editorconfig` file of our plugins, we use spaces for `.json` file. So I think it's fine to use spaces here.

AbdiTolesa replied:
I was just concerned about making the spacing consistent, either using tabs or spaces make it look cleaner.

truongwp commented on classes/frmBtsModApp.php:
```diff
@@ -62,14 +163,67 @@ private static function save_settings_for_footer( $atts, $content ) {
+			'<div class="wp-block-frm-modal-content',
+			'<div ' . self::get_modal_wrapper_attrs( $atts ) . ' class="modal fade wp-block-frm-modal-content',
+			$content
+		);
```
There are 2 replacements there. I updated the code to just use one replacement.

AbdiTolesa replied:
Oh, I see!

truongwp commented:
> @Crabcyborg @truongwp As a general comment and discussion point on this PR, it looks we are heavily relying on `__experimental ` features and it is [stated in this issue](https://github.com/WordPress/gutenberg/issues/40316) that every new Gutenberg release means potentially breaking changes. They don't recommend it for any serious production usage. I believe we should hardly use them or at least try to find as much alternative solutions as possible to minimise the maintenance burden required as new Gutenberg versions are released.
> 
> What are your thoughts on this?

There are some useful `__experimental`s like `__experimentalUseBorderProps`. I'm not sure what we should do here. We need to hear Mike's opinion.

Crabcyborg commented:
@truongwp I don't see alternative options.

According to https://make.wordpress.org/core/2022/08/10/proposal-stop-merging-experimental-apis-from-gutenberg-to-wordpress-core/ it sounds like it's likely to become part of Core as it's already included in Core and WordPress has a focus on maintaining backwards compatibility.

I think it may be worth trying to check if `__experimentalUseBorderProps` exists before accessing it, and checking if `useBorderProps` exists as well, in case it was made an official part of Core. But this is extra work that may not pay off. It's also likely to happen at some point in the future though as `__experimentalUseBorderProps` isn't a permanent long term name for the feature either.

Here's an issue about auditing the experimental API https://github.com/WordPress/gutenberg/issues/43635. There isn't anything here mentioning any immediate deprecations for anything.

We're using an `__unstable` function in Lite (`__unstableIsEditorReady`) and it has been fine so far. It's behind a `if ( 'function' !== typeof editor.__unstableIsEditorReady ) {` check just in case, so if it is removed, it won't result in errors.

Our Gutenberg blocks have also broken in the past, because official stable features were deprecated and removed, so the experimental and unsafe features are more of a best guess.

I'm okay leaving this decision with you @truongwp. If it does break, we just need to make sure we update it in a timely manner, so let's focus on something that's easy to update and patch in the future.

truongwp commented:
@Crabcyborg Thanks for finding that article.

> What about the existing experimental APIs?
> 
> Most of those already merged to Core would get a stable alias. It would preserve BC and shouldn’t noticeably affect the bundle size. Some will need a different treatment; let’s discuss that case-by-case.
> 
> What if an existing experimental API already in Core needs to be removed?
> 
> Does it? If so, let’s consider that on a case-by-case basis. There are established Core practices like contacting plugin authors, writing make Core posts, preferring soft deprecations, and so on. I don’t expect to see many instances of this.

I think it's ok to use experimental APIs which have been merged into Core.

truongwp commented:
@tuguirazvan Thank you. I didn't notice we use the same icon for our blocks. So I think using the same icon here will make sense. If we want different icons, that shouldn't be in this pull request scope.

stephywells commented:
This works great @truongwp ! Very nice!

I'm seeing a couple issues:
- I can't drag and drop a block into the modal before there are blocks added inside it. I can only use the + button. Is this an easy adjustment?
- Changing the modal size dropdown feels like it should change the preview size on the page. [In the design](https://www.figma.com/file/SXTkDpu190Du7HSk9SAJ6g/Modal-Add-On?type=design&node-id=203-20981&t=wGw51X65yC3WtBpO-0), it shows the overlay with the modal box inside it. Like this:
<img width="1306" alt="Screenshot 2023-05-30 at 5 27 37 PM" src="https://github.com/Strategy11/formidable-modal/assets/1116876/85f14e84-dd07-44a2-80e6-42ba57cca056">

- The overlay color setting is missing. Did we decide to wait on adding that?

truongwp commented:
@stephywells Oh. I don't know that we have a design for this.

I can drag a block to the empty modal content, but the placeholder styling is missing. It looks like an easy fix.

I will add the overlay and the color setting for it.

stephywells commented:
Nice update @truongwp ! I'll play around with this a bit more. From looking at the video, it looks like we could use a bit of margin between the button and modal box on the builder page. 

Also, is it possible to default the alpha value to something like 20% when setting the overlay color?

truongwp commented:
> Also, is it possible to default the alpha value to something like 20% when setting the overlay color?

Do you want to make it always 20%, or just the default color when the block is added?

If you want it's always 20%, I will disable the alpha input and process the color when changing.

If you want the default value, I'm not sure we can do it easily. I can set a default color with 20% alpha. But that color will be applied to the block.

stephywells commented:
> > Also, is it possible to default the alpha value to something like 20% when setting the overlay color?
> 
> Do you want to make it always 20%, or just the default color when the block is added?
> 
> If you want it's always 20%, I will disable the alpha input and process the color when changing.
> 
> If you want the default value, I'm not sure we can do it easily. I can set a default color with 20% alpha. But that color will be applied to the block.

It doesn't always need to be 20%. Maybe we can have the default color selected? Is there a smaller version of the color picker we can use like the colors in the button settings?

stephywells commented on formidable-modal.php:
```diff
@@ -1,29 +1,40 @@
+/**
+ * Plugin Name: Formidable Bootstrap Modal
+ * Description: Easily insert a link to open a form in a model
+ * Version: 2.1-beta
```
Let's make this v3.0

stephywells commented on blocks-src/frm-modal-content/editor.scss:
```diff
@@ -0,0 +1,60 @@
+	.modal-header {
+		display: flex;
+		flex-shrink: 0;
+		align-items: center;
```
It looks like this is in my theme and needs to be overridden (it moves the close icon out of the corner):
.modal-footer:after, .modal-footer:before, .modal-header:after, .modal-```
header:before {
    content: " ";
    display: table;
}
```

With previous version:
<img width="672" alt="Screenshot 2023-06-01 at 9 31 08 AM" src="https://github.com/Strategy11/formidable-modal/assets/1116876/883656e2-83a2-472e-8875-f0006a5ffa57">

With new version:
<img width="631" alt="Screenshot 2023-06-01 at 9 32 12 AM" src="https://github.com/Strategy11/formidable-modal/assets/1116876/44cd244c-31cd-41ef-b14c-3f7b3448d3a3">

truongwp replied:
I just hide them: https://github.com/Strategy11/formidable-modal/pull/16/commits/43a31eebe255265594a7e07c78ea34eb634f64da

Do you think this is the right solution?

stephywells replied:
Sure! That looks good to me now!

stephywells commented on classes/frmBtsModApp.php:
```diff
@@ -115,17 +288,17 @@ public static function output_modal() {
-				$modal .= '<a class="close frm_icon_font frm_cancel1_icon alignright" data-dismiss="modal" ></a>';
-				$modal .= '<h4 class="modal-title" id="frmModalLabel-' . esc_attr( $i ) . '">'. $title .'</h4>';
+				$modal .= '<h4 class="modal-title" id="frmModalLabel-' . esc_attr( $i ) . '">' . $title . '</h4>';
+				$modal .= '<a class="close frm_icon_font frm_cancel1_icon alignright" data-dismiss="modal" data-bs-dismiss="modal"></a>';
```
Can we switch to using a plan x that isn't in a circle? Here's the design for an example:
https://www.figma.com/file/SXTkDpu190Du7HSk9SAJ6g/Modal-Add-On?type=design&node-id=203-20981&t=GgfCjkQsAWqK5Bpa-0

truongwp replied:
I used the `&times;` icon. Please check if it's ok with you.

stephywells replied:
That's fine. We just need to make sure it doesn't use the link color, and that it's the right size, and uses the right placement (equal space from the top and side):
https://www.figma.com/file/SXTkDpu190Du7HSk9SAJ6g/Modal-Add-On?type=design&node-id=123-20441&t=W69Q5FIung59ca7c-0

stephywells replied:
Thanks! This is better. But my theme is still making the X blue with:
```
body a {
    color: #4199fd;
}
```

Can we set the color on the X?

stephywells replied:
@truongwp did you see this comment? I know it's a bit buried.

stephywells replied:
Looks good! Thank you!

stephywells commented on css/bootstrap-modal.css:
```diff
@@ -0,0 +1,464 @@
+	.modal-dialog {
+		max-width: var(--bs-modal-width);
+		margin-right: auto;
+		margin-left: auto;
```
The modal is closer to the top of the page now than it was before. This was margin: 30px auto; before.

Was this intentional?

truongwp replied:
This isn't intentional. Look like CSS of Bootstrap 5.3 causes this. I will override it in the plugin.

truongwp commented:
> Is there a smaller version of the color picker we can use like the colors in the button settings?

There isn't a component like that. That's a combination of some components: Button, ColorIndicator, Tabs,...

stephywells commented:
> Is it ok if the name of Overlay color setting is `Background Color`? That setting is perfect here, but I can't change the name to `Overlay Color`.

Well.. we have two settings in the design. Once for the modal color and one for the overlay color. I think that's fine to use Background color for the overlay. Is there a good option for the modal color?

tuguirazvan commented:
I found another icon design in the Figma files for the Modal Block. Will leave it here in case you want a third option.
![CleanShot 2023-06-02 at 12 05 40@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/3a12859e-c78b-483e-9ad6-3ff881ea00b9)

stephywells commented on blocks-src/frm-modal-content/helpers.js:
```diff
@@ -0,0 +1,15 @@
+
+export const ModalHeader = ( { title } ) => (
+	<div className="modal-header">
+		<h4 className="modal-title">{ title }</h4>
```
We decided to remove the hardcoded title, and let people add it themselves with blocks. This will be much more flexible if we don't have a title and border.
https://www.figma.com/file/SXTkDpu190Du7HSk9SAJ6g/Modal-Add-On?type=design&node-id=123-21163&t=W69Q5FIung59ca7c-0

truongwp replied:
I updated the block. I still keep the modal title input in the block edit. If it's empty, we won't show it in the frontend, and adjust a bit the position of the block button.

stephywells commented on css/bootstrap-modal.css:
```diff
@@ -0,0 +1,485 @@
+.modal {
+	--bs-modal-zindex: 1055;
+	--bs-modal-width: 500px;
+	--bs-modal-padding: 1rem;
```
Looks like we need to override this. It's 40px in our design.

stephywells replied:
This looks great now! Thank you!!

stephywells commented on blocks-src/frm-modal-content/edit.js:
```diff
@@ -0,0 +1,112 @@
+						<a href="#" onClick={ onClickResetOverlayColor }>{ __( 'Reset color', 'frmmodal' ) }</a>
+					</p>
+
+					<ColorPicker
```
Just moving the conversation here to make it easier to follow and resolve. Can we have the default color selected? 
Is there a smaller version of the color picker we can use like the colors in the button settings?

truongwp replied:
Yes, we can have a default color selected. But it will make the overlay color darker a bit. The color we apply here isn't the background of the `.modal-backdrop` element because Bootstrap uses just one `.modal-backdrop` for all modals (it's rgba(0,0,0,0.5) by default). I apply the Overlay color setting to the background of the modal. So the actual overlay color in the frontend will be the combination of rgba(0,0,0,0.5) and the Overlay color setting.

I'm creating a control for the small color picker.

truongwp replied:
@stephywells I completed this. Please take a look at those settings.

truongwp replied:
I'm working on the border of those color settings, and improving the code (move the setting code to a reusable component).

stephywells replied:
This works great now! Awesome!

It looks like we could add css to prevent the double overlay and making the colors darker than intended. Can we add `.modal-backdrop.show{ opacity:0 !important}`

Does that solve the problem for you?

truongwp replied:
@stephywells It will remove the backdrop of other modals (from other plugins).

stephywells replied:
What about this?
`.wp-block-frm-modal-content.show ~ .modal-backdrop.show{ opacity:0 !important}`

truongwp replied:
I have just thought about 1 solution. We can print a specific CSS code when opening the modal and remove it when closing (we'll need to use Bootstrap JS API). That CSS will change the color of the opening modal backdrop.

truongwp replied:
That works. Thank you.

truongwp commented:
@stephywells Is that 1 or 2?

> Here's the icon from the design that we should use:
https://www.figma.com/file/SXTkDpu190Du7HSk9SAJ6g/Modal-Add-On?type=design&node-id=203-20981&t=W69Q5FIung59ca7c-0

![Screen Shot 2023-06-05 at 10 45 19](https://github.com/Strategy11/formidable-modal/assets/19748318/ae1497f1-e07d-4b94-b426-271c0f5527e7)

tuguirazvan commented:
I think we should go for the FF logo as icon.

tuguirazvan commented:
Will leave my feedback on this one below: 

1. The close icon should be center aligned with the title of the modal. The icon should always stay between those red lines.
![CleanShot 2023-06-06 at 11 53 57@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/f5f060ec-b5ee-4634-9690-b6b73eecee2c)
![CleanShot 2023-06-06 at 11 55 47@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/70a2011d-8d14-42de-98de-50cde8f0568c)

2. Why do we lock the block by default? I don't think we should do that. _(I wanted to remove the block and didn't know what was happening because the "remove block" option wasn't available since we locked it by default)_
![CleanShot 2023-06-06 at 11 58 37@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/9a9dcf1d-9dbd-4701-8d2c-67d98791dcd7)

3. The paddings should stay consistent even for the inner elements. Below you can see that the paragraph block has more padding than the title and icon have. Then I'd increase the padding from 1em to 2em (top, bottom, left, right). For the title alone, we should keep 1em bottom padding. 
![CleanShot 2023-06-06 at 12 01 53@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/9ffe2904-d5f5-4435-8ed0-aee3da4769ea)
![CleanShot 2023-06-06 at 12 11 04@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/7aeb5d7f-deb7-400e-bed7-97645bc52f92)

4. Bold style applied to the title isn't working properly. The same applies to Italic.  
<img width="384" alt="CleanShot 2023-06-06 at 12 14 58@2x" src="https://github.com/Strategy11/formidable-modal/assets/8666554/fb2c700c-9341-4f9e-bf83-5ec61b9f7dfe">

Or it seems like we have it bold by default. What if our users want it in regular style? 

<img width="512" alt="CleanShot 2023-06-06 at 12 14 36@2x" src="https://github.com/Strategy11/formidable-modal/assets/8666554/dca325fd-805a-42a4-bc1a-9716542a89bf">

5. Can't we have different font sizes for the modal title and the inner content? When selecting a size, it affects the X close icon which shouldn't happen. The close icon should stay the same size all the time. 

![CleanShot 2023-06-06 at 12 58 56@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/54a7efdc-0541-476e-8696-1616d65eb0cc)
<img width="637" alt="CleanShot 2023-06-06 at 13 03 01@2x" src="https://github.com/Strategy11/formidable-modal/assets/8666554/6df67409-69f8-40c7-b797-f8d314cfdea3">

6. What if I want the modal to be centred on the page instead of pinned at the top of the page? If this is not on our roadmap for the first release, I would add a 16px or even 24px margin from the top. Having no margins at all it feels weird. 

7. Do we have plans on improving this modal block in the future? Because right now it lacks lots of customization options.

truongwp commented:
@tuguirazvan Thank you so much for the very detailed report.

For the 6th, I think we don't support centered vertical alignment. I'll add more top margin in case the admin bar showing. It's the reason you see the modal is stuck to the top.

For the 7th, we definitely have more customization options in the future. In this version, our main purpose is building current modals using block editor.

stephywells commented on blocks-src/frm-modal-content/edit.js:
```diff
@@ -0,0 +1,141 @@
+					<SelectControl
+						label={ __( 'Modal dialog size', 'frmmodal' ) }
+						options={ [
+							{ label: __( 'Default', 'frmmodal' ), value: '' },
```
With the 'Default' setting, it shows as 800px wide in the back-end and 500px wide on the page. When I switch between default and large, the preview size doesn't change.

stephywells replied:
The default should be 500px in the block instead of 800px.

stephywells commented on blocks-src/frm-modal-content/helpers.js:
```diff
@@ -0,0 +1,19 @@
+import { RichText } from '@wordpress/block-editor';
+
+export const CloseButton = () => (
+	<a className="close alignright" data-dismiss="modal" data-bs-dismiss="modal">&times;</a>
```
This still isn't aligning quite right for me. Also, I think we can remove alignright now too.
<img width="783" alt="Screenshot 2023-06-06 at 10 34 26 PM" src="https://github.com/Strategy11/formidable-modal/assets/1116876/bdd975e3-c70d-4fc5-89f2-f2b9fb7c8fa8">

truongwp replied:
I added a bit of margin-bottom to fix this. I think it's fine since the dimension of that button is fixed in px.

stephywells reviewed (changes requested):
This is super close! Great job on this @truongwp !

tuguirazvan commented:
@truongwp Thanks for doing all the changes. It looks awesome now, but I've found one more bug.


I think you still have extra padding at the bottom of the last block. See the examples below
| Image 1 | Image 2 |
|--------|--------|
| ![CleanShot 2023-06-07 at 16 35 13@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/015be87b-7de5-4174-9968-87681b71c411) | ![CleanShot 2023-06-07 at 16 37 08@2x](https://github.com/Strategy11/formidable-modal/assets/8666554/e89d010d-5870-44ab-822e-4f99713f8d25) |

truongwp commented:
@tuguirazvan That's because the last block has `margin-bottom: 1em;`. I reduced the `padding-bottom` of the modal body to `1em`. Thank you.

stephywells commented:
> That's because the last block has `margin-bottom: 1em;`. I reduced the `padding-bottom` of the modal body to `1em`. Thank you.

@truongwp Can we remove the padding from the last child instead to keep it consistent?
`.wp-block-frm-modal-content .modal-body > div > *:last-child{margin-bottom:0}`

truongwp commented:
> > That's because the last block has `margin-bottom: 1em;`. I reduced the `padding-bottom` of the modal body to `1em`. Thank you.
> 
> @truongwp Can we remove the padding from the last child instead to keep it consistent? `.wp-block-frm-modal-content .modal-body > div > *:last-child{margin-bottom:0}`

@stephywells Thank you. I updated it. I removed the top margin of the first block too. The margin setting in that block still works.

@tuguirazvan No problem at all. Let me check that.

tuguirazvan commented:
@truongwp @stephywells 
One more thing here. I'm able to click outside and close the modal. Do we want to enable this by default? Should we have an option for them to opt-in for this? 

Maybe we will run in some cases where I don't want my users to click outside and close the modal without taking the main action. Not sure if it makes sense now thb.

truongwp commented:
> @truongwp @stephywells One more thing here. I'm able to click outside and close the modal. Do we want to enable this by default? Should we have an option for them to opt-in for this?
> 
> Maybe we will run in some cases where I don't want my users to click outside and close the modal without taking the main action. Not sure if it makes sense now thb.

@tuguirazvan That should be the default behavior of the modal. We can add an option to disable that in the future. I think that option will be useful when using with the Trigger modal option in your design.

stephywells commented:
@truongwp The title in the modal is confusing. I think we should match the design and remove that completely from the block settings.

truongwp commented:
> @truongwp The title in the modal is confusing. I think we should match the design and remove that completely from the block settings.

Do we need a modal title block? Or users need to add a heading block, change the margin and border?

truongwp commented:
@stephywells I removed modal title completely.

stephywells reviewed (approved):
This looks ready to me! Thanks so much for all the updates @truongwp !

How do we deploy this to beta with the scripts compiled? It looks like the version on the QA site isn't up to date.

truongwp commented:
@stephywells Thanks for testing. I activated the correct plugin on the beta site.
